bigbone/src/main/kotlin/social/bigbone/api/Scope.kt

```diff
@@ -3,22 +3,257 @@ package social.bigbone.api
 /**
  * Represents the access permissions that can be requested when registering an app or when
  * requesting access tokens.
+ * @param scopes individual scopes that should be requested.
  */
-class Scope @JvmOverloads constructor(private vararg val scopes: Name = arrayOf(Name.ALL)) {
+class Scope(private vararg val scopes: Name) {
 
     /**
-     * The available scopes.
+     * The available scopes. These scopes are hierarchical, i.e. if you have access to [READ], you automatically have
+     * access to [READ_ACCOUNTS]. It is recommended that you request as little as possible for your application.
      */
     enum class Name(val scopeName: String) {
+        /**
+         * Grants access to read data. Requesting this scope will also grant its child scopes.
+         */
         READ("read"),
+
+        /**
+         * Child scope of [READ].
+         */
+        READ_ACCOUNTS("read:accounts"),
```

Comment:
How do you feel about using subclasses here instead so that we don’t have a long list of `Name` enums but would instead only have the top-level ones and you would then reference this one here as `Name.READ.ACCOUNTS`?

Reply:
I like that idea. Will see how to best achieve that without breaking too many things if at all possible. :)

Reply:
@PattaFeuFeu Making this work in Java as well wasn't trivial, but I think the end result is worth it even if there's some breaking change. All scopes can now be found hierarchically arranged in the Scope class, so that instead of, for example

`Scope(Scope.Name.READ, Scope.Name.ADMIN_READ_ACCOUNTS)`

it is now possible to write

`Scope(Scope.READ.ALL, Scope.ADMIN.READ.ACCOUNTS)`.

In addition to that, `Scope.Name` now is an interface that would allow library users to define and use new Mastodon scopes that we haven't implemented yet.

Can you have a look at whether that resolves this conversation?

Reply:
Interesting approach! I like it!

> In addition to that, Scope.Name now is an interface that would allow library users to define and use new Mastodon scopes that we haven't implemented yet.

I especially like that! 👍 

Suggestion time:

```kotlin
fun interface Name {
    fun name(): String
}
```

If we define the interface `Name` as a `fun interface` and turn the `val` into a `fun`, we can shorten the “implementations” quite a bit:

```kotlin
@JvmField
val ALL = Name { "read" }
```

If you like that approach, you can apply this git patch so that we won’t have to do the work twice. 😊 

[Replace_Name_interface_with_Name_fun_interface.patch](https://github.com/andregasser/bigbone/files/13679453/Replace_Name_interface_with_Name_fun_interface.patch)

---

I’m not the biggest fan of all those

```kotlin
/**
 * Child scope of [READ].
 */
```

kDocs. They don’t add much information. How about removing them and keeping only those that add some value, like

```kotlin
/**
 * Grants access to read all data. Requesting this scope is equivalent to requesting all other [READ] child scopes.
 */
```
?

If ktlint/detekt has any issue with that, I’d suggest to suppress those issues for this file / class.

Reply:
Making it a `fun interface` is perfect, thanks for that idea - and the patch. :)

Yes, I don't really like those kdoc, either. I was thinking about adding something along the lines of ", to be used in endpoints reading account data." to each one, but that would just be repeating the obvious. I'll remove them and suppress issues.